### PR TITLE
GetParentReferences should use namespace from RouteContext

### PR DESCRIPTION
**Pull request by zirain**
fix: https://github.com/envoyproxy/gateway/issues/3872

### CODE REVIEW

**Comment by arkodg** [comment]
is this a regression from https://github.com/envoyproxy/gateway/pull/3532 ?

**Comment by zirain** [comment]
> is this a regression from #3532 ?

looks like not

```
// GetParentReferences returns the ParentReference of the Route object.
func GetParentReferences(route RouteContext) []gwapiv1.ParentReference {
	rv := reflect.ValueOf(route).Elem()
	kind := rv.FieldByName("Kind").String()
	pr := rv.FieldByName("Spec").FieldByName("ParentRefs")
	if kind == KindHTTPRoute || kind == KindGRPCRoute {
		return pr.Interface().([]gwapiv1.ParentReference)
	}

	parentReferences := make([]gwapiv1.ParentReference, pr.Len())
	for i := 0; i < len(parentReferences); i++ {
		p := pr.Index(i).Interface().(gwapiv1.ParentReference)
		parentReferences[i] = UpgradeParentReference(p)
	}
	return parentReferences
}
```

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: internal/gatewayapi/testdata/httproute-with-multi-gateways-with-same-name.in.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          protocol: HTTP
+          port: 80
+          allowedRoutes:
+            namespaces:
+              from: All
+httpRoutes:
+  - apiVersion: gateway.networking.k8s.io/v1
+    kind: HTTPRoute
+    metadata:
+      namespace: default
```
Comment (arkodg) [inline internal/gatewayapi/testdata/httproute-with-multi-gateways-with-same-name.in.yaml:34]: shouldn't this be `envoy-gateway` ?

Comment (arkodg) [reply internal/gatewayapi/testdata/httproute-with-multi-gateways-with-same-name.in.yaml:34]: lets track this bug with a new GH issue

**Review by arkodg — APPROVED** [review]

**Comment by zirain** [comment]
hold for a while

**Comment by zirain** [comment]
@arkodg I think now we have the right fix.

File: internal/gatewayapi/helpers.go [hunk 2]
```diff
@@ -109,7 +109,7 @@ func KindDerefOr(kind *gwapiv1.Kind, defaultKind string) string {
 // to a Gateway with the given namespace/name, irrespective of whether a
 // section/listener name has been specified (i.e. a parent ref to a listener
 // on the specified gateway will return "true").
-func IsRefToGateway(parentRef gwapiv1.ParentReference, gateway types.NamespacedName) bool {
+func IsRefToGateway(parentRefNamespace gwapiv1.Namespace, parentRef gwapiv1.ParentReference, gateway types.NamespacedName) bool {
```
Comment (arkodg) [inline internal/gatewayapi/helpers.go:112]: should we rename `parentRefNamespace` to `routeNamespace` in this diff, so its clear the value is derived from the namespace of the route, not the parentRef itself (like a Gateway or Listener) ?

Comment (zirain) [reply internal/gatewayapi/helpers.go]: updated

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by shawnh2 — APPROVED** [review]

**Pull request merged**